### PR TITLE
New `optional` annotation on attributes

### PR DESCRIPTION
**Pull request by privat**
As requested at #1843 by @R4PaSs, a new annotation `optional` is now available on attributes.

~~~nit
class A
    var x: Int = 99 is optional
   # Because of `optional`, the automatic signature of the A constructor is `init(x: nullable Int)`
end

var a = new A
print a.x # outputs 99
var b = new A(4)
print b.x # output 4
~~~

In the model, the `optional` annotation only affects the signature and the behavior of the setter.
It transforms the argument to a `nullable` one and use the provided value if `null` is given as a parameter.

The `nullable` parameter is then propagated to the automatic constructors in the usual way.

Basically, the previous example is equivalent to having written:

~~~
class A
    var x: Int is noautoinit
    fun x=(x: nullable Int) is autoinit do if x != null then self.x = x else self.x = 99
end
~~~

### CODE REVIEW

File: tests/base_attr_optional.nit [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+a.i.output
+a.o.output
+
+a = new A(2)
+a.i.output
+a.o.output
+
+a = new A(3, true)
+a.i.output
+a.o.output
```
Comment (R4PaSs) [inline tests/base_attr_optional.nit]: Maybe an additional test for the optional named boolean could be useful ?

~~~nit
a = new A(o=true)
a.i.output
a.o.output
~~~

File: tests/base_attr_optional.nit [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+a.i = 1
+a.o = 10
+a.i.output
+a.o.output
+
+a.i = null
+a.o = null
```
Comment (R4PaSs) [inline tests/base_attr_optional.nit]: This behaviour I'm not sure of :/

I mean it makes sense, but is it really what we want ?
Should not the `i=` be protected in this case and inaccessible from this place ? Or maybe the `i=` should have another name for the initialization process ?

**Comment by R4PaSs** [comment]
I really like seeing this :)

Though there's the re-assignation issue which might surprise some people, though this could be fixed later if no real solution emerges soon.

:shipit: !

File: tests/base_attr_optional.nit (hunk 2 above)
Comment (privat) [inline tests/base_attr_optional.nit]: by default all writers are protected, so this should not leak unless explicitly stated by the programmer

Comment (privat) [inline tests/base_attr_optional.nit]: An alternative could be to distinguish the initializer and the setter. but thing bring an additional method.

**Comment by xymus** [comment]
+1
